packages/backend/src/apps/m365-excel/common/workbook-session/index.ts

```diff
@@ -0,0 +1,306 @@
+import type { IGlobalVariable } from '@plumber/types'
+
+import type { AxiosRequestConfig, AxiosResponse } from 'axios'
+
+import {
+  getM365TenantInfo,
+  type M365TenantInfo,
+} from '@/config/app-env-vars/m365'
+import HttpError from '@/errors/http'
+import logger from '@/helpers/logger'
+
+import { tryParseGraphApiError } from '../parse-graph-api-error'
+
+import {
+  clearSessionIdFromRedis,
+  getSessionIdFromRedis,
+  runWithLock,
+  setSessionIdInRedis,
+} from './redis'
+
+async function invalidateSessionId(
+  tenant: M365TenantInfo,
+  fileId: string,
+  badSessionId: string,
+): Promise<void> {
+  await runWithLock(tenant, fileId, async (signal) => {
+    // Nothing to do if another worker has already switched our fleet to a
+    // different session.
+    const sessionIdInRedis = await getSessionIdFromRedis(tenant, fileId)
+    if (sessionIdInRedis !== badSessionId) {
+      return
+    }
+
+    if (signal.aborted) {
+      throw new Error(`Redis lock error: ${signal.reason}`)
+    }
+
+    await clearSessionIdFromRedis(tenant, fileId)
+  })
+}
+
+async function refreshSessionId(
+  tenant: M365TenantInfo,
+  fileId: string,
+  $: IGlobalVariable,
+): Promise<string> {
+  return await runWithLock(tenant, fileId, async (signal) => {
```

Comment:
This will immediately fail the step; I will add retries in a later PR

Reply:
Is this what you are referring to? I thought session id expires in an hour?
<img width="904" alt="image" src="https://github.com/opengovsg/plumber/assets/65110268/30eac270-39e7-49d6-b104-7cd05a8fff94">

Reply:
does this get revalidated after an hour when the key expires?

Reply:
yes. Actually session ID expires in 5 minutes on Microsoft's end. Have changed the expiry timeout to match. I initially set expiry to an hour to ensure we would never start 2 sessions concurrently. But after discussion with ian, i think I was too paranoid.